commons/zenoh-shm/src/api/provider/shared_memory_provider.rs

```diff
@@ -511,6 +501,75 @@ unsafe impl<'a, Policy: AllocPolicy, IDSource, Backend: SharedMemoryProviderBack
     }
 }*/
 
+/// Builder for allocations
+#[zenoh_macros::unstable_doc]
+pub struct AllocBuilder2<
+    'a,
+    IDSource: ProtocolIDSource,
+    Backend: SharedMemoryProviderBackend,
+    Policy = JustAlloc,
+> {
+    data: AllocData<'a, IDSource, Backend>,
+    _phantom: PhantomData<Policy>,
+}
+
+// Generic impl
+impl<'a, IDSource, Backend, Policy> AllocBuilder2<'a, IDSource, Backend, Policy>
+where
+    IDSource: ProtocolIDSource,
+    Backend: SharedMemoryProviderBackend,
+{
+    /// Set the allocation policy
+    #[zenoh_macros::unstable_doc]
+    pub fn with_policy<OtherPolicy>(self) -> AllocBuilder2<'a, IDSource, Backend, OtherPolicy> {
+        AllocBuilder2 {
+            data: self.data,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+// Alloc policy
+impl<'a, IDSource, Backend, Policy> AllocBuilder2<'a, IDSource, Backend, Policy>
+where
+    IDSource: ProtocolIDSource,
+    Backend: SharedMemoryProviderBackend,
+    Policy: AllocPolicy,
+{
+    /// Get the result
+    #[zenoh_macros::unstable_doc]
+    pub fn res(self) -> BufLayoutAllocResult {
+        let layout = AllocLayout::new(self.data).map_err(ZLayoutAllocError::Layout)?;
+
+        layout
+            .alloc()
+            .with_policy::<Policy>()
+            .res()
+            .map_err(ZLayoutAllocError::Alloc)
+    }
+}
+
+// Async Alloc policy
+impl<'a, IDSource, Backend, Policy> AllocBuilder2<'a, IDSource, Backend, Policy>
+where
+    IDSource: ProtocolIDSource + Send + Sync,
+    Backend: SharedMemoryProviderBackend + Sync,
+    Policy: AsyncAllocPolicy,
+{
+    /// Get the async result
+    #[zenoh_macros::unstable_doc]
+    pub async fn res_async(self) -> BufLayoutAllocResult {
```

Comment:
Same comment as above. We should use either `.await` or `.wait()` by implementing `Wait` or `IntoFuture` traits.